File: prototype/rails/src/app/domain/listings/listing_availability.rb
require_relative "listing_status"

module Domain
  module Listings
    module ListingAvailability
      # A sold listing keeps its page so the links a buyer already followed
      # still lead somewhere; a draft or archived one was never public.
      ON_STOREFRONT = [ListingStatus::FOR_SALE, ListingStatus::SOLD].freeze

      module_function

      def on_storefront?(status)
        ON_STOREFRONT.include?(status)
      end

      def purchasable?(status, quantity)
        status == ListingStatus::FOR_SALE && quantity.positive?
      end
    end
  end
end

File: prototype/node/src/app/core/listings/listing-availability.ts
import type { ListingStatus } from './listing-status.ts'

// A sold listing keeps its page so a link a buyer already followed still
// leads somewhere; a draft or archived one was never public.
export const STOREFRONT_STATUSES: readonly ListingStatus[] = ['for_sale', 'sold']

/** What the grid and the search results draw: a sold piece keeps its page and leaves the grid. */
export const BROWSABLE_STATUSES: readonly ListingStatus[] = ['for_sale']

export function isOnStorefront(status: ListingStatus, hasActiveRemoval: boolean): boolean {
  return !hasActiveRemoval && STOREFRONT_STATUSES.includes(status)
}

export function isPurchasable(status: ListingStatus, quantity: number, hasActiveRemoval: boolean): boolean {
  return !hasActiveRemoval && status === 'for_sale' && quantity > 0
}
